src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs

```diff
@@ -251,106 +366,198 @@ bool ShouldAnalyze(SymbolStartAnalysisContext context, INamedTypeSymbol namedTyp
+                    return false;
 
-        // Property and field have to agree on type.
-        if (!property.Type.Equals(getterField.Type))
-            return;
+                // If the user made the field readonly, we only want to convert it to a property if we
+                // can keep it readonly.
+                if (getterField.IsReadOnly && !@this.SupportsReadOnlyProperties(compilation))
+                    return false;
 
-        // Mutable value type fields are mutable unless they are marked read-only
-        if (!getterField.IsReadOnly && getterField.Type.IsMutableValueType() != false)
-            return;
+                // Mutable value type fields are mutable unless they are marked read-only
+                if (!getterField.IsReadOnly && getterField.Type.IsMutableValueType() != false)
+                    return false;
 
-        // Don't want to remove constants and volatile fields.
-        if (getterField.IsConst || getterField.IsVolatile)
-            return;
+                // Field and property have to be in the same type.
+                if (!containingType.Equals(getterField.ContainingType))
+                    return false;
 
-        // Field and property should match in static-ness
-        if (getterField.IsStatic != property.IsStatic)
-            return;
+                // Field and property should match in static-ness
+                if (getterField.IsStatic != property.IsStatic)
+                    return false;
+
+                // Property and field have to agree on type.
+                if (!property.Type.Equals(getterField.Type))
+                    return false;
 
-        var fieldReference = getterField.DeclaringSyntaxReferences[0];
-        if (fieldReference.GetSyntax(cancellationToken) is not TVariableDeclarator { Parent.Parent: TFieldDeclaration fieldDeclaration } variableDeclarator)
+                if (!TryGetSyntax(getterField, out _, out var variableDeclarator, cancellationToken))
+                    return false;
+
+                var initializer = @this.GetFieldInitializer(variableDeclarator, cancellationToken);
+                if (initializer != null && !@this.SupportsPropertyInitializer(compilation))
+                    return false;
+
+                if (!@this.CanConvert(property))
+                    return false;
+
+                // Can't remove the field if it has attributes on it.
+                var attributes = getterField.GetAttributes();
+                if (attributes.Length > 0 && !@this.SupportsFieldAttributesOnProperties)
+                    return false;
+
+                return true;
+            },
+            (this, compilation, containingType, property, cancellationToken));
+
+        if (getterFields.IsEmpty)
             return;
 
+        var isTrivialSetAccessor = false;
+
         // A setter is optional though.
-        var setMethod = property.SetMethod;
-        if (setMethod != null)
+        if (property.SetMethod != null)
```

Comment:
similar to above.  if there is a set-method, we now do a deeper analysis to see which fields it uses.